### ISSUE TITLE
forward_pageant module outdated

### ISSUE THREAD

**Issue by WhyNotPal** [rapid7/metasploit-framework#15046]
## Steps to reproduce

Recommended of windows/manage/forward_pageant  (tested on msf6): 
msf  exploit(handler) > use post/windows/manage/forward_pageant 
msf  post(forward_pageant) > set SESSION 1
msf  post(forward_pageant) > run

## Were you following a specific guide/tutorial or reading documentation?

See https://labs.f-secure.com/archive/abusing-putty-and-pageant-through-native-functionality/
and related repo: https://github.com/rapid7/metasploit-framework/pull/5380 

## Expected behavior

Should run 

## Current behavior

Crashes with message: Post failed: NoMethodError undefined method `make_tmpname' for Dir::Tmpname:Module

### Metasploit version

msf6

## Additional Information

This module relies on the 'make_tmpname' method that is removed from Ruby since 2.5.0. 
See https://medium.com/@maciejmensfeld/ruby-2-5-0-upgrade-remarks-100d7c0dd73d

Another way to generate a temp filename should be defined/used.

**Comment by WhyNotPal** [comment]
The code for 'make_tmpname' still exists there: 
https://apidock.com/ruby/Dir/Tmpname/make_tmpname
So it's relatively easy to re-create a function to the same effect. However, the reason why this function was deprecated in Ruby 2.5.0 is not mentioned, so I don't know if it's a good idea.

**Comment by WhyNotPal** [comment]
That being said, there might be more to it. This module was written in 2015/6 and even after re-defining the outdated function, I can't get it to work as it should. There are issues in the interface with the pageant - 
ssh-add -l -> "invalid format error"
ssh (remote server) -> does not pass the pageant : "Unsuccessful response received"

Pity - it must be a great tool if it works.

**Comment by bcoles** [comment]
> That being said, there might be more to it. This module was written in 2015/6 and even after re-defining the outdated function, I can't get it to work as it should. There are issues in the interface with the pageant - ssh-add -l -> "invalid format error" ssh (remote server) -> does not pass the pageant : "Unsuccessful response received"
> 
> Pity - it must be a great tool if it works.

After fixing the temp file creation this module works for me, but it's a bit flakey. It works eventually after trying multiple times.

Tested on Windows 7 SP1 (x64).

```
[2022-08-07 12:37:38] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:39] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:40] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:40] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:40] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:41] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:42] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:42] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:43] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:43] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:44] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:44] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:44] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:45] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:45] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:45] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:46] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:46] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:46] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:47] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
error fetching identities: agent refused operation
[2022-08-07 12:37:47] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/ivqeiqnh" ssh-add -l
3072 SHA256:/M07p51CmCSMrV1lbFs19OMvyRw6g9Wxbq8bW5px0KA asdf@ubuntu-22-04-amd64 (RSA)
[2022-08-07 12:37:47] root@kali:~/Desktop/metasploit-framework# 
```

```
[2022-08-07 12:41:53] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
Welcome to Ubuntu 22.04 LTS (GNU/Linux 5.15.0-25-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

209 updates can be applied immediately.
29 of these updates are standard security updates.
To see these additional updates run: apt list --upgradable

*** System restart required ***
Last login: Sun Aug  7 12:41:32 2022 from 192.168.200.130
asdf@ubuntu-22-04-amd64:~$ exit
logout
Connection to 192.168.200.158 closed.
[2022-08-07 12:42:03] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:06] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:09] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
Welcome to Ubuntu 22.04 LTS (GNU/Linux 5.15.0-25-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

209 updates can be applied immediately.
29 of these updates are standard security updates.
To see these additional updates run: apt list --upgradable

*** System restart required ***
Last login: Sun Aug  7 12:42:27 2022 from 192.168.200.130
asdf@ubuntu-22-04-amd64:~$ 
logout
Connection to 192.168.200.158 closed.
[2022-08-07 12:42:13] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:17] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:19] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:21] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:23] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
asdf@192.168.200.158's password: 

[2022-08-07 12:42:24] root@kali:~/Desktop/metasploit-framework# SSH_AUTH_SOCK="/tmp/yszfbfew" ssh asdf@192.168.200.158
Welcome to Ubuntu 22.04 LTS (GNU/Linux 5.15.0-25-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

209 updates can be applied immediately.
29 of these updates are standard security updates.
To see these additional updates run: apt list --upgradable

*** System restart required ***
Last login: Sun Aug  7 12:42:37 2022 from 192.168.200.130
asdf@ubuntu-22-04-amd64:~$ 
```

The module verbose output offers a hint, in that when a very small reply is received the command will fail:

```
[*] PageantJacker: Received data from socket (size: 4)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 1)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 4)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 1)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 4)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 1)
[*] PageantJacker: Response received (Success='true' Size='5' Error='0: No error')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='447' Error='0: No error')
```

**Comment by bcoles** [comment]
The temporary file issue was resolved in #16881.

As for the socket flakiness, I'm not sure whether this is an issue with `ssh`, pageant, or the Metasploit implementation (probably).

Given that it took over a year for someone to look at this issue, it is unlikely that anyone is going to look at the flakiness issue. The module works well, but sometimes needs a few attempts to establish a connection.

Closing this issue, but feel free to reopen if the module does not work at all for you, or if someone is keen to look into it.
